week2/part4/tensor.hpp

```diff
@@ -44,59 +44,251 @@ class Tensor
 
+    // Print the tensor elements
+    out << "Tensor Elements:" << std::endl;
+    for (size_t i = 0; i < shape[0]; ++i)
+    {
+        for (size_t j = 0; j < shape[1]; ++j)
+        {
+            for (size_t k = 0; k < shape[2]; ++k)
+            {
+                // Assuming a 3D tensor, adjust as needed for other ranks
+                out << "Tensor[" << i << "][" << j << "][" << k << "] = " << tensor({i, j, k}) << std::endl;
+            }
+        }
+    }
+
+    return out;
 }
 
 // Reads a tensor from file.
-template< Arithmetic ComponentType >
-Tensor< ComponentType > readTensorFromFile(const std::string& filename)
+template <Arithmetic ComponentType>
+Tensor<ComponentType> readTensorFromFile(const std::string& filename)
 {
-    // TODO: Implement this function to read in tensors from file.
-    //       The format is defined in the instructions.
-
+    // TODO: Implement this function
 }
 
+// This is WIP, not yet working
+// Explanation (UKL): The function doesn't write anything to the file, but it doesn't throw any errors either.
 // Writes a tensor to file.
-template< Arithmetic ComponentType >
-void writeTensorToFile(const Tensor< ComponentType >& tensor, const std::string& filename)
+template <Arithmetic ComponentType>
+void writeTensorToFile(const Tensor<ComponentType>& tensor, const std::string& filename)
 {
-    // TODO: Implement this function to write tensors to file.
-    //       The format is defined in the instructions.
+    // Open the file in binary mode for writing
+    std::ofstream outFile(filename, std::ios::binary);
+
+    if (!outFile.is_open())
+    {
+        // Handle error: unable to open the file
+        std::cerr << "Error: Unable to open file for writing: " << filename << std::endl;
+        return;
+    }
+
+    // Write the rank of the tensor to the file
+    size_t rank = tensor.rank();
+    outFile.write(reinterpret_cast<const char*>(&rank), sizeof(size_t));
+    if (outFile.fail()) {
+        std::cerr << "Error: Failed to write rank to the file." << std::endl;
+        outFile.close();  // Close the file before returning
+        return;
+    }
+
+    // Write the shape of the tensor to the file
+    const std::vector<size_t> &shape = tensor.shape();
+    outFile.write(reinterpret_cast<const char *>(shape.data()), rank * sizeof(size_t));
+    if (outFile.fail()) {
+        std::cerr << "Error: Failed to write shape to the file." << std::endl;
+        outFile.close();  // Close the file before returning
+        return;
+    }
+
+    // Write the tensor elements to the file
+    const std::vector<ComponentType> &data = tensor.getData();
+    outFile.write(reinterpret_cast<const char *>(data.data()), data.size() * sizeof(ComponentType));
+    if (outFile.fail()) {
+        std::cerr << "Error: Failed to write tensor elements to the file." << std::endl;
+        outFile.close();  // Close the file before returning
+        return;
+    }
 
+    // Close the file
+    outFile.close();
 }
```

Comment:
This is a WIP; As indicated in the code, the function is not able to create and write to a file. No visible errors are outputted either.